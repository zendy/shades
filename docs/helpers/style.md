# `style` (Pseudo-Selectors, Attributes and Props)

The `style` helper is a DSL for handling pseudo-selectors, attributes, props, and includes many combinators for each.  It supports all known pseudo-classes, pseudo-elements and pseudo-functions, as well as many advanced features for selecting against attributes and their values.  Lastly, it includes special handlers for doing pattern-matching against props (React-only support for now).

Example:

```js
import React from 'react';
import shades from '@bupa-digital/shades/react';
import style from '@bupa-digital/shades/utils.style';

// A random collection of different style features
const PseudoIcon = shades.i({
  [style.element.before]: {
    // ::before
    fontSize: '15px',
    content: 'Hello there!'
  },
  [style.hover]: {
    // :hover
    [style.element.after]: {
      // :hover::after
      fontFamily: 'Material Icons',
      content: 'close'
      [style.prop.dark]: {
        color: 'white'
      },
      [style.props.all(style.prop.dark, style.prop.theme)]: ({ theme }) => ({
        background: theme.dark
      })
    }
  },
  [style.attr.title]: {
    // [title]
    fontSize: '20px'
  },
  [style.and(style.attr.href.startsWith('https'), style.attr.href.endsWithAny('jpg', 'png'))]: {
    // So its an image
    background: ({ href }) => `url(${href |> JSON.stringify})`
  }
});

// ... later that same day ...
return (
  <PseudoIcon theme={mySpecialTheme} dark href="https://example.com/img.png" />
);
```

## Pseudo-Selectors

There are 3 types of pseudo-selector:
1. pseudo-classes, like `hover`, `active` and `visited`
2. pseudo-elements, like `before`, `after` and `selection`
3. pseudo-functions, like `nth-child`, `not` and `lang`

You can find the list of supported pseudo-selectors [over here](#supported-pseudo-selectors)

You can use any of the supported pseudo-classes and pseudo-functions by name on the style object, like so:

```js
style.hover // :hover
style.focus // :focus
style.not(style.active) // :not(:active)
```

For pseudo-elements, you can access them also by name via `style.element`:

```js
style.element.before // ::before
style.element.after // ::after
```

Not the most exciting example, to be honest.  But where it really gets powerful is in using combinators (coming up in the next section)

### Escape-hatches

In case there are pseudo-selectors not yet supported out of the box, we've provided some escape hatches so you can define non-supported pseudos:

```js
style.pseudo('just-a-test') // :just-a-test
style.pseudo('custom-pseudo-function', 'hello') // :custom-pseudo-function(hello)
style.element('wow-amazing') // ::wow-amazing
```

### Supported Pseudo-Selectors:

I opted to include all pseudo selectors that I could.  A few were left out deliberately, such as `host` and `host-context`, but please raise an issue if you need them or if we're missing any others.

#### Pseudo-Classes

- active
- any-link
- checked
- default
- disabled
- empty
- enabled
- first
- first-child
- first-of-type
- fullscreen
- focus
- hover
- indeterminate
- in-range
- invalid
- last-child
- last-of-type
- left
- link
- only-child
- only-of-type
- optional
- out-of-range
- read-only
- read-write
- required
- right
- root
- scope
- target
- valid
- visited

#### Pseudo-Functions

- any
- dir
- lang
- matches
- not
- nth-child
- nth-last-child
- nth-last-of-type
- nth-of-type

#### Pseudo-Elements

- before
- after
- backdrop
- cue
- first-letter
- first-line
- grammar-error
- placeholder
- selection
- spelling-error

## Combinators

Combinators are functions that combine more than one of something in particular ways.  Say for example you wanted a selector that would activate when the element was hovered and focussed at the same time.  Again, not the most exciting thing in the world:

```js
style.and(style.hover, style.focus) // :hover:focus
```

Or if the element is either hovered OR focussed:

```js
style.or(style.hover, style.focus) // :hover, :focus
```

But what about a situation whereby you wanted a more complicated selector, perhaps compounded combinators?

```js
style.and(style.hover, style.or(style.visited, style.active, style.not(style.onlyChild)))
// :hover:visited, :hover:active, :hover:not(:only-child)
```

That's all well and good, but it's still more to type than the CSS version - until we get to the next section and look at attributes!

The supported combinators are:
- `and` (which combines)
- `or` (which comma-separates)

> Combinators can be used with pseudo-selectors and attributes - however they **cannot** be used with props at this time, which have their own special combinators.

## Attributes

Since attributes can often be a mix of both specification-based and custom/arbitrary, we decided not to restrict the supported attributes to any predefined list.  Instead, you can select for any attribute name you like, both as simple presence selectors (`[title]`) and value expressions (`[title$="example"]`) with a nice DSL.

Some examples:

```js
style.attr.title // [title]
```

Simple presence selecctor.  If the title attribute is passed to the Shades element, this will activate.  Ya know, normal CSS stuff.  But then comes...

```js
style.attr.href.startsWithAny('http', 'https', 'ftp')
// [href^="http"], [href^="https"], [href^="ftp"]
```

Ok, now that's gotta be easier to type than the CSS version.

You can mix and match attribute and pseudo-selectors however you wish, just as you would in CSS:

```js
style.not(style.attr.href('localhost'))
// :not([href="localhost"])
style.and(style.attr.contains('example'), style.not(style.hover))
// [attr*="example"]:not(:hover)
```

Note that this will **not work** yet:

```js
style.not(style.and(style.hover, style.active)) // BREAKS SILENTLY
```

> Attempting to use combinators inside a pseudo-function **will break and not even warn you**, but it's on the list of things to make work.

One last thing, theres also built in support for data attributes, and they support everything that normal attributes do:

```js
style.data.specialValidation
// [data-special-validation]
style.data.specialValidation.anyOf('email', 'phone', 'date')
// [data-special-validation="email"], [data-special-validation="phone"], [data-special-validation="date"]
```

#### Supported Attribute Combinators

To select for an attribute with an exact value, you can just call the attribute like a method:
`style.attr.myAttribute('my value')` becomes `[my-attribute="my value"]`

The following methods and combinators are supported for attribute helpers:

- `anyOf(...values)` - select any of these exact values for the given attribute
- `contains(value)` - select only if the attribute contains the given value
- `containsAny(...values)` - select if any of the given values is found in the attribute value
- `startsWith(value)` - select if the attribute value starts with the given value
- `startsWithAny(...values)` - select if the attribute value starts with any of the given values
- `endsWith(value)` - select if the attribute value ends with the given value
- `endsWithAny(...values)` - select if the attribute value ends with any of the given values

## Properties and Pattern Matching

Property matching, also known as pattern matching, can be done with a similar syntax to attribute selectors, although there is not yet support for the various attribute combinators/value matching features of attribute selectors.  Example:

```js
style.prop.fancyGreetings
```

Although one thing you can do with property matching that you can't with all other features is the ability to specify a function that takes the property value as its argument, like so:

```js
const MoneyButton = shades.button({
  // Here, the `value` argument is set to the value of the `currencyType` property if it is provided
  [style.prop.currencyType]: (value) => value === 'dollars' && ({
    [style.element.before]: {
      // Whereas all other functions still receive all props as normal
      content: ({ amount }) => `Only $${amount}.99`
    }
  })
})

// A few lines later...

return (
  <MoneyButton currencyType="dollars" amount="39" />
)
```

You can also match against multiple properties, with the `any` and `all` combinators.  You can acccess the combinators via `props` rather than `prop`.  Example:

```js
style.props.any(style.prop.dollars, style.prop.euros)
// Will match if any of the given props are defined, in this case, dollars or euros
style.props.all(style.prop.currencyType, style.prop.amount)
// Will match only when all the given props are defined, in this ase. currencyType and amount
```

> Note: When using a function with a `props` coombinator, the function will receive all props, rather than just the target ones.

> Note: You **cannot** mix property matchers in pseudo/attribute combinators, i.e `style.any(style.hover, style.prop.fooBar)` - doing so could irreparably damage the fabric of reality and transform all physical matter into cosmic nothingness.
