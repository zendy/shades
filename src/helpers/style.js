import {
  compose,
  map,
  concat,
  join,
  curry,
  prop,
  propEq,
  equals,
  mapObjIndexed
} from 'ramda';

import styleStore, { getDescriptor } from './style/selector-store';

import {
  dasherize,
  isString,
  isSymbol,
  isNumber,
  isArray,
  isDefined,
  when,
  betterSet,
  includes,
  stateful,
  proxyFunction,
  proxyPropertyGetter,
  proxyRecord,
  anyOf,
  allOf,
  firstItem,
  lastItem
} from '../utilities';

import {
  PSEUDO_SELECTORS,
  COMBINATORS,
  COMBINATOR_INSERTS
} from './selector-types'

const isDescriptorSym    = Symbol('Compute Selector');
const isDescriptor       = (value) => value?.[isDescriptorSym] ?? false;

const keyFromSymbol      = (...args) => Symbol.keyFor(...args);
const symbolFromKey      = (...args) => Symbol.for(...args);
const styleStoreKey = Symbol.for('Shades: styleStore');

const asPseudoClass      = (name) => `:${name |> dasherize}`;
const asPseudoElement    = (name) => `::${name |> dasherize}`;
const asPropertySelector = (givenName) => `!!${givenName}`;

const asPseudoFunction = curry((name, value) => (
  `:${name |> dasherize}(${value})`
));

const storeDescriptor = (descriptorItem) => {
  const symbolKey = descriptorItem.symbolKey;

  styleStore.lift(
    ({ addItem }) => addItem(symbolKey, descriptorItem)
  );

  return descriptorItem;
};

const createDescriptor = (kind, config = {}) => (value) => {
  const keyOrValue = (
    config?.stringKey ?? value
    |> when(isString).otherwise(JSON.stringify)
  );

  const symbolKey  = symbolFromKey(keyOrValue);

  const selfDescriptor = {
    [isDescriptorSym]:    true,
    kind,
    value,
    symbolKey,
    originalKey:          keyOrValue,
    // this toString doesnt actually return a string
    // it instead returns a symbol whose key is a string
    toString:        () => symbolKey
  };

  return selfDescriptor;
};

const toString = (value) => value.toString();

const descriptorToString = when(isDescriptor).then(prop('originalKey'));
const anythingToString = when(isString).otherwise(toString);

const createAndStoreDescriptor = (kind, config) => compose(
  storeDescriptor,
  createDescriptor(kind, config)
);

const createCombinator = (kind) => (...data) => {
  const stringKey = (
    data
    |> map(
      compose(
        anythingToString,
        descriptorToString
      )
    )
    |> join(` ${COMBINATOR_INSERTS[kind]} `)
  );

  return data |> createAndStoreDescriptor(
    kind,
    { stringKey }
  );
}

const pseudoCombinators = {
  and: createCombinator(
    COMBINATORS.COMBINATOR_AND
  ),
  or:  createCombinator(
    COMBINATORS.COMBINATOR_OR
  )
};

const withAttribute = (givenName) => {
  const quoteString = JSON.stringify;

  const attrWithValue  = (givenValue) => `[${givenName}=${givenValue  |> quoteString}]`;
  const attrStartsWith = (givenValue) => `[${givenName}^=${givenValue |> quoteString}]`;
  const attrEndsWith   = (givenValue) => `[${givenName}$=${givenValue |> quoteString}]`;
  const attrContains   = (givenValue) => `[${givenName}*=${givenValue |> quoteString}]`;

  const anyCombinator = (mapperFn) => (...givenValues) => pseudoCombinators.or(
    ...givenValues.map(mapperFn)
  );

  const plainValue = `[${givenName}]`;

  const outerMethods = {
    anyOf:         anyCombinator(attrWithValue),
    contains:      attrContains,
    containsAny:   anyCombinator(attrContains),
    startsWith:    attrStartsWith,
    startsWithAny: anyCombinator(attrStartsWith),
    endsWith:      attrEndsWith,
    endsWithAny:   anyCombinator(attrEndsWith),
    toString: () => plainValue
  };

  return proxyFunction(
    attrWithValue,
    outerMethods
  )
}

const pseudoClassHandler = (specialChains) => (targetName) => {
  if (PSEUDO_SELECTORS.CLASSES.includes(targetName)) return (
    targetName |> asPseudoClass
  );

  if (PSEUDO_SELECTORS.FUNCTIONS.includes(targetName)) return (
    compose(
      asPseudoFunction(targetName),
      descriptorToString
    )
  );

  if (PSEUDO_SELECTORS.ELEMENTS.includes(targetName)) return (
    targetName |> asPseudoElement
  )

  return specialChains?.[targetName];
}



const style = (
  proxyPropertyGetter(
    pseudoClassHandler({
      ...pseudoCombinators,
      element: proxyPropertyGetter(asPseudoElement, asPseudoElement),
      pseudo: (name, value) => {
        if (isDefined(value)) return asPseudoFunction(name, value);
        return asPseudoClass(name);
      },
      data: proxyPropertyGetter(
        when(isString).then(compose(
          withAttribute,
          concat('data-'),
          dasherize
        ))
      ),
      attr: proxyPropertyGetter(
        when(isString).then(compose(
          withAttribute,
          dasherize
        ))
      ),
      prop: proxyPropertyGetter(
        when(isString).then(asPropertySelector)
      ),
      props: ({
        any: createCombinator(COMBINATORS.PROPERTY_OR),
        all: createCombinator(COMBINATORS.PROPERTY_AND)
      })
    })
  )
)

export default style;
