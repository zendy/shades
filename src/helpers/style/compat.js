import {
  createStyleHelpers
} from './style-helper';

export default createStyleHelpers({ useProxyFeatures: false });
