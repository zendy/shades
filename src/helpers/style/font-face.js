import {
  compose,
  map,
  concat,
  join,
  curry,
  prop,
  propEq,
  equals,
  mapObjIndexed,
  apply
} from 'ramda';

import {
  dasherize,
  isString,
  isSymbol,
  isNumber,
  isArray,
  isDefined,
  when,
  betterSet,
  includes,
  proxyPropertyGetter,
  anyOf,
  allOf,
  firstItem,
  lastItem
} from '../utilities';

// const fontFaceStore = stateful({
//   index: 1,
//   cached: new Map()
// }, {
//   addToCache: ({ cached, index }, hashKey, parsedValue) => ({
//     cached: cached.set(hashKey, parsedValue),
//     index: index + 1
//   })
// });

const fontFacePlainKeys = ['src', 'unicodeRange'];
const isTuple           = allOf(isArray, propEq('length', 2));
const isLocal           = allOf(isTuple, compose(equals('local'), firstItem));
const asLocal           = (value) => `local(${value |> JSON.stringify})`;
const asFormat          = (value) => `format(${value |> JSON.stringify})`;
const asUrl             = (value) => `url(${value |> JSON.stringify})`;

const asUrlWithFormat = (format, url) => (
  [
    url    |> asUrl,
    format |> asFormat
  ] |> join(' ')
);

export const parseSrc = (items) => when(isString).otherwise(
  map(
    when(isTuple).then(
      when(isLocal)
        .then(lastItem, asLocal)
        .otherwise(apply(asUrlWithFormat))
    )
  )
);



//
// const createFontFace = (fontName, fontItem) => {
//   return mapObjIndexed((value, key) => {
//     const newKey = key |> when(includedIn(fontFacePlainKeys)).otherwise(dasherize, concat('font-'));
//
//
//   })
// }
//
// const fontFaceHandler = (...givenItems) => {
//   const fromCache    = fontFaceStore.getState('cached');
//   const currentIndex = fontFaceStore.getState('index');
//
//   return givenItems |> when(fromCache.has).then(fromCache.get).otherwise(
//     (argumentValues) => {
//       const familyName = (
//         argumentValues
//         |> firstItem
//         |> when(isString).otherwise(
//           concat('shades-fontface-', currentIndex)
//         )
//       );
//
//       const
//     }
//   )
// }
