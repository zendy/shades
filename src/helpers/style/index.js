import {
  createStyleHelpers
} from './style-helper';

export default createStyleHelpers({ useProxyFeatures: true });
