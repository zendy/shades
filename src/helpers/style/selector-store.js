import {
  prop
} from 'ramda';

import {
  stateful,
  when,
  isSymbol
} from '../../utilities';

const storeValueGlobally = (name) => (value) => {
  const scope = global ?? window;

  if (scope?.[name]) return scope[name];

  scope[name] = value;

  return value;
}

const styleStoreKey = Symbol.for('Shades: style selector store');

const styleStore = stateful(
  new Map(),
  {
    addItem: (store, itemKey, itemValue) => store.set(itemKey, itemValue)
  }
) |> storeValueGlobally(styleStoreKey);

export const getDescriptor = (key) => {
  const stateKey = key |> when(isSymbol).otherwise(prop('symbolKey'));
  const output = styleStore.getState(stateKey);

  if (!output) console.error('Could not find that key in the store for some reason', { key, stateKey, store: styleStore.getState(), output })

  return output;
}

export default styleStore;
