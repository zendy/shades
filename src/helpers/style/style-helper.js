import {
  compose,
  map,
  concat,
  join,
  curry,
  prop,
  propEq,
  pipe,
  equals,
  mapObjIndexed,
  has,
  both
} from 'ramda';

import selectorRegistry, {
  asPropertySelector,
  isDescriptorSelector,
  isDescriptorObject,
  createDescriptor
} from '../../registries/selectors';

import {
  dasherize,
  isString,
  isSymbol,
  isNumber,
  isArray,
  isDefined,
  when,
  getPath,
  includes,
  proxyCatchAll,
  anyOf,
  allOf,
  firstItem,
  lastItem,
  toString
} from '../../utilities';

import {
  PSEUDO_SELECTORS,
  COMBINATORS,
  COMBINATOR_INSERTS,
} from '../selector-types'

const keyFromSymbol      = (...args) => Symbol.keyFor(...args);
const symbolFromKey      = (...args) => Symbol.for(...args);

const asPseudoClass      = (name) => `:${name |> dasherize}`;
const asPseudoElement    = (name) => `::${name |> dasherize}`;

const asPseudoFunction = curry((name, value) => (
  `:${name |> dasherize}(${value})`
));

const descriptorObjectToString = when(isDescriptorObject).then(prop('key'));

const anythingToString = (
  when(both(isSymbol, selectorRegistry.hasDescriptor)).then(
    selectorRegistry.getDescriptor,
    prop('key')
  ).otherwise(
    descriptorObjectToString.otherwise(toString)
  )
)

const createAndStoreDescriptor = (kind, itemKey) => compose(
  selectorRegistry.addDescriptor,
  createDescriptor(kind, itemKey)
);

const createCombinator = (kind) => (...data) => {
  const stringifiedData = data |> map(anythingToString);
  const computedKey = (
    stringifiedData
    |> join(` ${COMBINATOR_INSERTS[kind]} `)
  );

  return stringifiedData |> createAndStoreDescriptor(
    kind,
    computedKey
  );
}

const joinSelectorsWith = (separator) => (data) => (
  data |> map(anythingToString) |> join(` ${separator} `)
);

const joiners = {
  all: joinSelectorsWith(COMBINATOR_INSERTS[COMBINATORS.COMBINATOR_AND]),
  any: joinSelectorsWith(COMBINATOR_INSERTS[COMBINATORS.COMBINATOR_OR])
}

const pseudoCombinators = {
  and: (...data) => (
    selectorRegistry.helpers.selectors.all(
      data,
      joiners.all(data)
    )
  ),
  or: (...data) => (
    selectorRegistry.helpers.selectors.any(
      data,
      joiners.any(data)
    )
  )
};

const asAttributeSelector = (original) => {
  const givenName = original |> dasherize;
  const quoteString = JSON.stringify;

  const attrWithValue  = (givenValue) => `[${givenName}=${givenValue  |> quoteString}]`;
  const attrStartsWith = (givenValue) => `[${givenName}^=${givenValue |> quoteString}]`;
  const attrEndsWith   = (givenValue) => `[${givenName}$=${givenValue |> quoteString}]`;
  const attrContains   = (givenValue) => `[${givenName}*=${givenValue |> quoteString}]`;

  const anyCombinator = (mapperFn) => (...givenValues) => pseudoCombinators.or(
    ...givenValues.map(mapperFn)
  );

  const plainValue = `[${givenName}]`;

  const outerMethods = {
    equals:        attrWithValue,
    anyOf:         anyCombinator(attrWithValue),
    contains:      attrContains,
    containsAny:   anyCombinator(attrContains),
    startsWith:    attrStartsWith,
    startsWithAny: anyCombinator(attrStartsWith),
    endsWith:      attrEndsWith,
    endsWithAny:   anyCombinator(attrEndsWith),
    toString: () => plainValue
  };

  return outerMethods;
}

const asDataAttributeSelector = compose(
  asAttributeSelector,
  concat('data-')
)

const createPropsWith = (valueCreator) => (keyList) => (
  keyList.reduce((result, key) => ({
    ...result,
    [key]: valueCreator(key)
  }), {})
);

const allPseudoClasses = PSEUDO_SELECTORS.CLASSES |> createPropsWith(asPseudoClass)
const allPseudoElements = PSEUDO_SELECTORS.ELEMENTS |> createPropsWith(asPseudoElement);
const allPseudoFunctions = PSEUDO_SELECTORS.FUNCTIONS |> createPropsWith(asPseudoFunction);

export const createStyleHelpers = ({ useProxyFeatures = false } = {}) => {
  const withPropCatcher = when(useProxyFeatures).then(proxyCatchAll)

  return {
    ...pseudoCombinators,
    ...allPseudoClasses,
    ...allPseudoFunctions,

    pseudo: (name, value) => {
      if (isDefined(value)) return asPseudoFunction(name, value);
      return asPseudoClass(name);
    },

    element: allPseudoElements,
    elementOf: asPseudoElement, // escape-hatch

    attr: asAttributeSelector |> withPropCatcher,
    data: asDataAttributeSelector |> withPropCatcher,
    prop: asPropertySelector |> withPropCatcher,

    selector: (value) => selectorRegistry.helpers.selectors.arbitrary(value),

    props: {
      any: createCombinator(COMBINATORS.PROPERTY_OR),
      all: createCombinator(COMBINATORS.PROPERTY_AND)
    }
  }
}
