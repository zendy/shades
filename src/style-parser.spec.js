import css, {
  parseAllStyles,
  parseAndStringify
} from './style-parser';

import {
  states,
  mq
} from './helpers';

import style from './helpers/style';

const parseRulesNoDebug = parseAllStyles
const parseRulesWithDebug = parseAllStyles

const topSelector = '#meow';

describe('parseRules', () => {
  it('takes an object of rules and gives me back some more rules as strings', () => {
    const result = parseRulesNoDebug(topSelector, {}, {
      fontSize: '10px',
      color: 'blue',
      fontWeight: 'normal'
    });

    expect(result).toHaveProperty([topSelector]);
  })
  it('combines nested pseudo-elements and pseudo-selectors with the main selector', () => {
    const topSelectorHover = `${topSelector}:hover`;
    const topSelectorBefore = `${topSelector}::before`;

    const result = parseRulesNoDebug(topSelector, {}, {
      fontSize: '10px',
      color: 'blue',
      ':hover': {
        color: 'red'
      },
      '::before': {
        content: 'hi there',
        border: '1px solid #000'
      },
      fontWeight: 'normal'
    });

    expect(result).toHaveProperty([topSelector]);
    expect(result).toHaveProperty([topSelectorHover]);
    expect(result).toHaveProperty([topSelectorBefore]);

    expect(result[topSelector]).toMatchObject({
      'font-size': '10px',
      'color': 'blue',
      'font-weight': 'normal'
    });
    expect(result[topSelectorHover]).toMatchObject({
      'color': 'red'
    });
    expect(result[topSelectorBefore]).toMatchObject({
      'content': '"hi there"',
      'border': '1px solid #000'
    });
  })
  it('supports the style combinators', () => {
    const topSelectorHoverFocus = `${topSelector}:hover, ${topSelector}:focus`;

    const result = parseRulesNoDebug(topSelector, {}, {
      fontSize: '10px',
      color: 'blue',
      [style.or(style.hover, style.focus)]: {
        color: 'purple',
        border: '1px solid #fff'
      }
    });

    expect(result).toHaveProperty([topSelectorHoverFocus]);

    const expectedStyles = {
      'color': 'purple',
      'border': '1px solid #fff'
    };

    expect(result[topSelectorHoverFocus]).toMatchObject(
      expectedStyles
    );
  })
  // TODO: stop using this as my "test-everything-ever" bucket
  it('should shift at-rules up to the top level', () => {
    const topSelectorHover = `${topSelector}:hover`;
    const mq = '@media screen and (max-width: 800px)';

    const result = parseRulesNoDebug(topSelector, { }, {
      fontSize: '10px',
      color: 'blue',
      [mq]: {
        fontSize: '90px',
        background: 'green',
        ':hover': {
          border: '1px solid red',
          fontWeight: '800'
        }
      },
      textShadow: ['1px 1px 1px black', '2px 3px 4px white'],
      bubbles: ({dots}) => `1px ${dots ? 'dotted' : 'solid'} black`,
      ':hover': ({ cheese }) => ({
        fontFamily: cheese ? 'CheeseyMono' : 'MilkMono',
        border: `${cheese || 'none'} 100px`
      })
    });

    expect(result).toHaveProperty([topSelector]);
    expect(result).toHaveProperty([mq, topSelector]);
    expect(result).toHaveProperty([mq, topSelectorHover]);

    expect(result[topSelector]).toMatchObject({
      'font-size': '10px',
      'color': 'blue',
      'bubbles': '1px solid black'
    });

    expect(result[mq][topSelector]).toMatchObject({
      'font-size': '90px',
      'background': 'green'
    });
    expect(result[mq][topSelectorHover]).toMatchObject({
      'font-weight': '800',
      'border': '1px solid red'
    });
  });

  it('supports using functions as rules that take all given props as its argument', () => {
    const result = parseRulesNoDebug(
      topSelector,
      { dark: true, darkColor: 'navy' },
      {
        fontWeight: 'normal',
        color: (allProps) => allProps.darkColor ?? 'fallback'
      }
    );

    expect(result[topSelector]).toMatchObject({
      'font-weight': 'normal',
      'color': 'navy'
    })
  });

  it('really seriously supports all possible permutations of functions as style rule values', () => {
    const resultSimple = parseRulesNoDebug(
      topSelector,
      { dark: 'blue' },
      {
        [mq().portrait().from(370)]: (allProps) => ({
          [style.prop.dark]: value => ({
            color: ({ dark }) => value === dark && allProps.dark === value && 'navy'
          })
        })
      }
    );

    expect(resultSimple[mq().portrait().from(370)][topSelector]).toMatchObject({
      'color': 'navy'
    });
  });



  describe('Pattern matching', () => {
    it('supports pattern-matching rules for props', () => {
      const result = parseRulesNoDebug(topSelector, { dark: true }, {
        fontSize: '10px',
        color: 'green',
        [style.prop.dark]: {
          color: 'navy'
        },
        fontWeight: 'normal'
      });

      expect(result).toHaveProperty([topSelector]);
      expect(result[topSelector]).toMatchObject({
        'color': 'navy'
      })
    })

    it('supports a default value when no pattern match found', () => {
      const result = parseRulesNoDebug(topSelector, { }, {
        fontSize: '10px',
        color: 'green',
        [style.prop.dark]: {
          color: 'navy'
        },
        fontWeight: 'normal'
      });

      expect(result).toHaveProperty([topSelector]);
      expect(result[topSelector]).toMatchObject({
        'color': 'green'
      })
    })

    it('supports a pattern match rule as a function that takes the value of the named prop', () => {
      const result = parseRulesNoDebug(topSelector, { mode: 'dark' }, {
        fontSize: '10px',
        color: 'blue',
        [style.prop.mode]: value => ({
          color: `green-${value}`,
        }),
        fontWeight: 'normal'
      });


      expect(result).toHaveProperty([topSelector]);
      expect(result[topSelector]).toMatchObject({
        'font-size': '10px',
        'color': 'green-dark',
        'font-weight': 'normal'
      });
    })

    it('uses the default value if the matched function returns null, undefined, or false', () => {
      const result = parseRulesNoDebug(topSelector, { mode: 'supersayan' }, {
        fontSize: '10px',
        color: 'green',
        [style.prop.mode]: (value) => ({
          color: value === 'dark' && 'navy',
          padding: '100px'
        }),
        fontWeight: 'normal'
      });

      expect(result).toHaveProperty([topSelector]);
      expect(result[topSelector]).toMatchObject({
        'font-size': '10px',
        'color': 'green',
        'padding': '100px',
        'font-weight': 'normal'
      });
    })

    it('skips the rule entirely if there is no match', () => {
      const result = parseRulesNoDebug(topSelector, { mode: 'something unknown', nextOne: 'hello' }, {
        fontSize: '10px',
        [style.prop.mode]: (value) => ({
          color: value === 'dark' && 'navy'
        }),
        [style.prop.kittensEverywhere]: {
          color: 'purple'
        },
        fontWeight: 'normal'
      });

      expect(result).toHaveProperty([topSelector]);
      expect(result[topSelector]).not.toHaveProperty(['color']);
    })

    it('will render a block of styles for a block pattern', () => {
      const result = parseRulesNoDebug(
        topSelector,
        { mode: 'hi there', nextOne: 'dodgerblue' },
        {
          [style.prop.mode]: {
            fontWeight: 'bold',
            color: 'purple'
          },
          [style.prop.nextOne]: (value) => ({
            color: value,
            border: '1px solid #ccc'
          })
        }
      );

      expect(result[topSelector]).toMatchObject({
        'font-weight': 'bold',
        'color': 'dodgerblue',
        'border': '1px solid #ccc'
      });
    });

    it('renders style.element helpers as pseudo-elements', () => {
      const topSelectorBefore = `${topSelector}::before`;
      const topSelectorAfter = `${topSelector}::after`;

      const result = parseRulesNoDebug(
        topSelector,
        { dark: true },
        {
          color: 'blue',
          [style.element.before]: {
            fontWeight: 'bold',
            color: 'purple'
          },
          [style.element.after]: {
            [style.prop.dark]: {
              color: 'red'
            },
            [style.prop.light]: {
              color: 'green'
            }
          }
        }
      );

      const expectedOutput = {
        [topSelector]: { 'color': 'blue' },
        [topSelectorBefore]: { 'font-weight': 'bold', 'color': 'purple' },
        [topSelectorAfter]: { 'color': 'red' },
      };

      expect(result).toMatchObject(expectedOutput);
    });

    it('supports pseudo-functions from the style helper', () => {
      const topSelectorNotHover = `${topSelector}:not(:hover)`;
      const topSelectorHref = `[href^="http"]`;

      const result = parseRulesNoDebug(
        topSelector,
        { dark: true },
        {
          color: 'blue',
          [style.not(style.hover)]: {
            fontWeight: 'bold',
            color: 'purple'
          },
          [style.attr.href.startsWithAny('http', 'https')]: {
            [style.nthOfType('even')]: {
              [style.prop.dark]: {
                color: 'red'
              },
              [style.prop.light]: {
                color: 'green'
              }
            }
          }
        }
      );

      console.log(
        result
      )

      const expectedOutput = {
        [topSelector]: {
          'color': 'blue'
        },
        [`${topSelector}[href^="http"]:nth-of-type(even), ${topSelector}[href^="https"]:nth-of-type(even)`]: {
          'color': 'red'
        },
      };

      expect(result).toMatchObject(expectedOutput);
    });

  })
})

describe('parseAndStringify', () => {
  it('should not fail spectacularly', () => {
    const fakeGeneratedClass = 'lol-what-1234';
    const fakeProps = { mode: 'something', nextOne: 'hello' };

    const result = parseAndStringify(fakeGeneratedClass, fakeProps, {
      fontSize: '10px',
      color: {
        mode: (value) => value + '-alright',
        kittensEverywhere: 'purple'
      },
      fontWeight: 'normal'
    });

    expect(result).toEqual([
      `${fakeGeneratedClass} { font-size: 10px;color: something-alright;font-weight: normal; }`
    ])
  });

});
