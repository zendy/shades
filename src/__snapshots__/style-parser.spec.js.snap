// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyRules should not fail spectacularly 1`] = `
Array [
  "lol-what-1234:hover:active::before,lol-what-1234:focus:active::before,lol-what-1234:hover:visited::before,lol-what-1234:focus:visited::before,lol-what-1234:hover:active::after,lol-what-1234:focus:active::after,lol-what-1234:hover:visited::after,lol-what-1234:focus:visited::after { font-weight: super-ultra-bold; }",
  "@media screen and (min-width:500px) { lol-what-1234:hover:active::before,lol-what-1234:focus:active::before,lol-what-1234:hover:visited::before,lol-what-1234:focus:visited::before,lol-what-1234:hover:active::after,lol-what-1234:focus:active::after,lol-what-1234:hover:visited::after,lol-what-1234:focus:visited::after { font-family: helvetica-is-trendy; } }",
  "@media screen and (min-width:500px) { lol-what-1234 { fake-attribute: whats-up; } }",
  "lol-what-1234:hover:active,lol-what-1234:focus:active,lol-what-1234:hover:visited,lol-what-1234:focus:visited { font-family: whatever; }",
  "lol-what-1234 { font-size: 10px;color: something-alright;background: something-alright;font-weight: normal; }",
  "lol-what-1234:hover { background: red; }",
  "lol-what-1234:hover { text-decoration: underline; }",
]
`;

exports[`stringifyRules should use autoprefixer to generate vendor prefixes 1`] = `
Array [
  ".prefix-me { display: -ms-grid; display: grid;-webkit-user-select: none;-moz-user-select: none;-ms-user-select: none;user-select: none; }",
]
`;
